storm-core/src/jvm/org/apache/storm/dependency/DependencyUploader.java

```diff
@@ -0,0 +1,166 @@
+            blobStore = Utils.getClientBlobStore(conf);
+        }
+    }
+
+    public void shutdown() {
+        if (blobStore != null) {
+            blobStore.shutdown();
+        }
+    }
+
+    @VisibleForTesting
+    void setBlobStore(ClientBlobStore blobStore) {
+        this.blobStore = blobStore;
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<String> uploadFiles(List<File> dependencies, boolean cleanupIfFails) throws IOException, AuthorizationException {
+        checkFilesExist(dependencies);
+
+        List<String> keys = new ArrayList<>(dependencies.size());
+        try {
+            for (File dependency : dependencies) {
+                String fileName = dependency.getName();
+                String key = BlobStoreUtils.generateDependencyBlobKey(BlobStoreUtils.applyUUIDToFileName(fileName));
+
+                try {
+                    uploadDependencyToBlobStore(key, dependency);
+                } catch (KeyAlreadyExistsException e) {
+                    // it should never happened since we apply UUID
+                    throw new RuntimeException(e);
+                }
+
+                keys.add(key);
+            }
+        } catch (Throwable e) {
+            if (blobStore != null && cleanupIfFails) {
+                deleteBlobs(keys);
+            }
+            throw new RuntimeException(e);
+        }
+
+        return keys;
+    }
+
+    public List<String> uploadArtifacts(Map<String, File> artifacts) {
+        checkFilesExist(artifacts.values());
+
+        List<String> keys = new ArrayList<>(artifacts.size());
+        try {
+            for (Map.Entry<String, File> artifactToFile : artifacts.entrySet()) {
+                String artifact = artifactToFile.getKey();
+                File dependency = artifactToFile.getValue();
+
+                String key = BlobStoreUtils.generateDependencyBlobKey(convertArtifactToJarFileName(artifact));
+                try {
+                    uploadDependencyToBlobStore(key, dependency);
+                } catch (KeyAlreadyExistsException e) {
+                    // we lose the race, but it doesn't matter
+                }
+
+                keys.add(key);
+            }
+        } catch (Throwable e) {
+            throw new RuntimeException(e);
+        }
+
+        return keys;
+    }
+
+    private String convertArtifactToJarFileName(String artifact) {
+        return artifact.replace(":", "-") + ".jar";
+    }
+
+    private boolean uploadDependencyToBlobStore(String key, File dependency)
+            throws KeyAlreadyExistsException, AuthorizationException, IOException {
+
+        boolean uploadNew = false;
+        try {
+            // FIXME: we can filter by listKeys() with local blobstore when STORM-1986 is going to be resolved
+            // as a workaround, we call getBlobMeta() for all keys
+            blobStore.getBlobMeta(key);
+        } catch (KeyNotFoundException e) {
+            // TODO: do we want to add ACL here?
```

Comment:
Before that, we need to check that nimbus and supervisor can access these blobs. I didn't look deeply with blobstore so I would like someone to confirm this.
@revans2 Could you guide this thing?